GCC 10.2 maybe-unitialized failure

While trying to publish [libbinaryen to the OPAM repository](https://github.com/ocaml/opam-repository/pull/22337#issuecomment-1286556941), I encountered a bunch of failures on their Debian 11 CI. It seems that Debian 11 ships with GCC 10.2 which seems to have `-Wmaybe-unitialized` enabled by default (I can't actually find release notes for that change). It's possible this is enabled by `-Wall`

I've currently added `-Wno-maybe-unitialized` to our CMake flags, but ideally Binaryen would build under GCC 10 without any extra flags.

Here's a snippet of the error log from the OPAM CI:

```log

#=== ERROR while compiling libbinaryen.109.0.0 ================================#
# context              2.2.0~alpha~dev | linux/x86_64 | ocaml-base-compiler.4.14.0 | pinned(https://github.com/grain-lang/libbinaryen/releases/download/v109.0.0/libbinaryen-v109.0.0.tar.gz)
# path                 ~/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0
# command              ~/.opam/opam-init/hooks/sandbox.sh build dune build -p libbinaryen -j 31
# exit-code            1
# env-file             ~/.opam/log/libbinaryen-7-ef1bcd.env
# output-file          ~/.opam/log/libbinaryen-7-ef1bcd.out
### output ###
[...]
# -- Build files have been written to: /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen
# File "dune", line 78, characters 0-582:
#  78 | (rule
#  79 |  (target dllbinaryen.so)
#  80 |  (locks binaryen)
# .....
# 103 |      -DCMAKE_INSTALL_PREFIX=binaryen)
# 104 |     (run cmake --build binaryen --config Release -- -j4)
# 105 |     (copy binaryen/lib/libbinaryen.so dllbinaryen.so)))))
# (cd _build/default && /usr/bin/cmake --build binaryen --config Release -- -j4)
[...]
# [ 66%] Building CXX object src/passes/CMakeFiles/passes.dir/GenerateDynCalls.cpp.o
# [ 67%] Building CXX object src/passes/CMakeFiles/passes.dir/GlobalRefining.cpp.o
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp: In function 'wasm::WATParser::Result<typename Ctx::TypeT> wasm::WATParser::{anonymous}::valtype(Ctx&, wasm::WATParser::{anonymous}::ParseInput&) [with Ctx = wasm::WATParser::{anonymous}::ParseModuleTypesCtx]':
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp:179:13: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
#   179 |     return {};
#       |             ^
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp: In function 'wasm::WATParser::Result<typename Ctx::TypeT> wasm::WATParser::{anonymous}::valtype(Ctx&, wasm::WATParser::{anonymous}::ParseInput&) [with Ctx = wasm::WATParser::{anonymous}::ParseTypeDefsCtx]':
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp:179:13: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
#   179 |     return {};
#       |             ^
# At global scope:
# cc1plus: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
# cc1plus: note: unrecognized command-line option '-Wno-implicit-int-float-conversion' may have been intended to silence earlier diagnostics
# cc1plus: all warnings being treated as errors
# gmake[2]: *** [src/wasm/CMakeFiles/wasm.dir/build.make:251: src/wasm/CMakeFiles/wasm.dir/wat-parser.cpp.o] Error 1
# gmake[1]: *** [CMakeFiles/Makefile2:497: src/wasm/CMakeFiles/wasm.dir/all] Error 2
# gmake[1]: *** Waiting for unfinished jobs....
[...]
# gmake: *** [Makefile:149: all] Error 2
# (cd _build/default && /usr/bin/cmake -S binaryen -B binaryen -G 'Unix Makefiles' -DCMAKE_CXX_FLAGS=-Wno-unused-variable -DBUILD_TESTS=OFF -DBUILD_TOOLS=OFF -DBUILD_STATIC_LIB=ON -DCMAKE_BUILD_TYPE=Release -DCMAKE_INSTALL_PREFIX=binaryen)
# -- Building with -DBUILD_LLVM_DWARF
# -- Building with -Wall
# -- Building with -Werror
# -- Building with -Wextra
# -- Building with -Wno-unused-parameter
# -- Building with -fno-omit-frame-pointer
# -- Building with -fno-rtti
# -- Building with -Wno-implicit-int-float-conversion
# -- Building with -Wno-unknown-warning-option
# -- Building with -Wswitch
# -- Building with -Wimplicit-fallthrough
# -- Building with -Wnon-virtual-dtor
# -- Building with -fPIC
# -- Building with -O2
# -- Building with -UNDEBUG
# -- Building libbinaryen as statically linked library.
# -- Configuring done
# -- Generating done
# -- Build files have been written to: /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen
# File "dune", line 26, characters 0-506:
# 26 | (rule
# 27 |  (targets libbinaryen.a)
# 28 |  (locks binaryen)
# ....
# 47 |      -DCMAKE_INSTALL_PREFIX=binaryen)
# 48 |     (run cmake --build binaryen --config Release -- -j4)
# 49 |     (copy binaryen/lib/libbinaryen.a libbinaryen.a)))))
# (cd _build/default && /usr/bin/cmake --build binaryen --config Release -- -j4)
[...]
# [ 93%] Building CXX object src/wasm/CMakeFiles/wasm.dir/wat-parser.cpp.o
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp: In function 'wasm::WATParser::Result<typename Ctx::TypeT> wasm::WATParser::{anonymous}::valtype(Ctx&, wasm::WATParser::{anonymous}::ParseInput&) [with Ctx = wasm::WATParser::{anonymous}::ParseModuleTypesCtx]':
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp:179:13: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
#   179 |     return {};
#       |             ^
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp: In function 'wasm::WATParser::Result<typename Ctx::TypeT> wasm::WATParser::{anonymous}::valtype(Ctx&, wasm::WATParser::{anonymous}::ParseInput&) [with Ctx = wasm::WATParser::{anonymous}::ParseTypeDefsCtx]':
# /home/opam/.opam/4.14/.opam-switch/build/libbinaryen.109.0.0/_build/default/binaryen/src/wasm/wat-parser.cpp:179:13: error: '<anonymous>' may be used uninitialized in this function [-Werror=maybe-uninitialized]
#   179 |     return {};
#       |             ^
# At global scope:
# cc1plus: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
# cc1plus: note: unrecognized command-line option '-Wno-implicit-int-float-conversion' may have been intended to silence earlier diagnostics
# cc1plus: all warnings being treated as errors
# gmake[2]: *** [src/wasm/CMakeFiles/wasm.dir/build.make:251: src/wasm/CMakeFiles/wasm.dir/wat-parser.cpp.o] Error 1
# gmake[1]: *** [CMakeFiles/Makefile2:497: src/wasm/CMakeFiles/wasm.dir/all] Error 2
# gmake: *** [Makefile:149: all] Error 2
```

--- a/src/wasm/wat-parser.cpp
+++ b/src/wasm/wat-parser.cpp
@@ -224,7 +224,7 @@ struct ParseInput {
         return n;
       }
     }
-    return {};
+    return std::nullopt;
   }
 
   std::optional<int64_t> takeS64() {
@@ -254,7 +254,7 @@ struct ParseInput {
         return n;
       }
     }
-    return {};
+    return std::nullopt;
   }
 
   std::optional<int32_t> takeS32() {
@@ -296,7 +296,7 @@ struct ParseInput {
         return d;
       }
     }
-    return {};
+    return std::nullopt;
   }
 
   std::optional<float> takeF32() {
@@ -306,7 +306,7 @@ struct ParseInput {
         return f;
       }
     }
-    return {};
+    return std::nullopt;
   }
 
   std::optional<std::string_view> takeString() {

